### ISSUE TITLE
[Question/Debate/Discuss] hard linewrapping without losing source-based diff ability

### ISSUE THREAD

**Issue by krishnakumarg1984** [lervag/vimtex#1317]
This is _not_ a bug report or not even specific to `vimtex`. Sorry about that. I am also pretty sure that a couple of older issues discussing this same aspect exists (but I could not find them by my search phrases in github issues).

This is intended to be a general discussion on a topic that has irked me for a really long time now - _prose diffing_ . Obviously, please feel free to close this. I only raise this as an "issue" here since there is a higher likelihood of many like-minded people sharing this thought.

I really like my source code formatted with hard line-wrapped to 80 characters width. I even like the flush margin since it gives a sense of neatness, i.e. I justify the text.  In `vim`, I set the `formatprg` to `par \ -jw80`  which uses the excellent `par` formatter written by Adam Costello. Other options such as Perl's `Text::Autoformat` might be equally powerful as well.   I find the justification feature can actually be used since `latex` ignores multiple spaces between words, and therefore does not affect the final PDF output.

However, as almost everyone here knows, this sort of wrapping messes up `diff`ing of the source code. 

The suggested workaround is to use soft linewrapping and putting one sentence per line. In my opinion, this is a slightly ungainly solution.  There is some mental cognitive workload and discipline needed to do this for a large document, especially when working with collaborators.  The full "paragraph-like" look for paragraphs is also adversely affected with a sentence ending abruptly just 23 characters into the line and new sentences beginning.  This gives a rather ugly/ragged look to the source paragraphs.

It is 2019 and there should be some natural language based diff tools that can handle this, right? I did some investigation, and the two closest things that I could find were:

* [Pandiff](https://github.com/davidar/pandiff)
* [difference.rs](https://github.com/johannhof/difference.rs)
* [jsdiff](https://github.com/kpdecker/jsdiff)

`difference.rs` and `pandiff` certainly look promising to me, and maybe `vimtex` can integrate it somehow?  Anyway, I'd like to ask if other users here have better recommendations or are willing to share their thoughts on this.

**Comment by Rmano** [comment]
@krishnakumarg1984  I commented the thing a lot of time ago [on StackExchange](https://vi.stackexchange.com/questions/6004/text-formatting-cvs-friendly), and referred the very interesting idea [by Christopher Lemmer Webber](http://dustycloud.org/blog/vcs-friendly-patchable-document-line-wrapping/)... but I do not know if this is worthwhile, really.

**Comment by andymass** [comment]
@krishnakumarg1984 , are you aware of `git diff --ignore-all-space` and `diff -b`/`diff -w`?  Also look at vim's `diffopt` settings.

**Comment by krishnakumarg1984** [comment]
@andymass  Yes, but they don't work when linebreaks change between the compared source files, like in a paragraph.  Essentially, all the standard `diff` tools are line-_oriented_, even if they can narrow the diff down to the word-level.

What we need is a paragraph-_oriented_ tool i.e. treat the paragraph as a whole without caring about line-breaks, resolve the diffs down to word-level granularity, and finally present a unified `diff` interface with the additions/removals to the document author.

**Comment by krishnakumarg1984** [comment]
@Rmano  The idea by Christoph Webber is a good workaround, but in my opinion, hampers readability of the paragraph.  Based on semantics and what's currently taught to me by librarians at my university, "A paragraph must convey a single cohesive idea.".  In striving to keep up with VCS compatibility, the ragged-lines in the source puts an extra cognitive load on the document writer.

"If you do enough work in any sort of free software environment, you get used
  to doing lots of writing of documentation or all sorts of other things in
  some plaintext system which exports to some non-plaintext system.
One way or another you have to decide: are you going to wrap your lines with
  newlines?
And of course the answer should be "yes" because lines that trail all the way
  off the edge of your terminal is a sin against the plaintext gods, who are
  deceptively mighty, and whose wrath is to be feared (and blessings to be
  embraced).
So okay, of course one line per paragraph is off the table.
So what do you do?"

That's just so hard to parse mentally. To me they appear as a collection of many separate sentences, and I have to force myself to think this as a paragraph conveying a single idea.

**Comment by krishnakumarg1984** [comment]
@lervag thank you. Can you try the difference.rs rust binary? Maybe this is the right tool?

**Comment by lervag** [comment]
@krishnakumarg1984 Based on the documentation, it seems to be a relatively general tool to find difference between two texts. This won't work for LaTeX, because we want to diff both the text and the code. We need a diff tool that is somehow syntax aware and that can ignore some whitespace but not all. I seems to me to be quite challenging to solve this particular problem.

**Comment by andymass** [comment]
@krishnakumarg1984 , you can combine `git diff --ignore-all-space` with `--word-diff`.

There is also the `wdiff` and `dwdiff` utilities.

Nothing smart or understanding of natural language, but they all work at the word-level.

**Comment by krishnakumarg1984** [comment]
@lervag I think that [`jsdiff`](https://github.com/kpdecker/jsdiff) meets the requirements, doesn't it.

It is based on the research paper published [here](http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.4.6927)

Look at this!

>
Diff.diffWords(oldStr, newStr[, options]) - diffs two blocks of text, comparing word by word, ignoring whitespace.
Returns a list of change objects (See below).
Options
ignoreCase: Same as in diffChars. 


Isn't that what we want exactly?

**Comment by krishnakumarg1984** [comment]
@andymass Wow. I didn't know about this. Does this work well in practise?

Among all the 4 choices you mentioned, which permutations/combinations do you use for your `git diff`
settings in your `.git/config` ?  Which is the best tool? Can you provide a 1-line summary of the tools you introduced in this discussion?

**Comment by krishnakumarg1984** [comment]
Apparently, google research has been working on something similar. Do you guys think this Is this useful?

[diff-match-patch](https://github.com/google/diff-match-patch)

**Comment by lervag** [comment]
@krishnakumarg1984 I don't understand how to combine this with vimtex. vimtex does not have any diffing features. I think this should be added/adopted into some other project. In any case, could you explain exactly how you would like this to work? That is, can you outline the workflow, including mappings/commands, that utilizes the possibly external tools?

**Comment by krishnakumarg1984** [comment]
@lervag This is my proposed workflow for a hard-wrapped/justified workflow.

1.  Have some mapping like `<leader>ld` that is active only when it is detected that the project uses `git`. 

2. Analogous to `\lt`  which brings up a split buffer containing the table of contents, the `\ld` mapping should bring up a `diff` buffer populated with output of a suitable external diff command (discussed below) that the document's author can quickly refer to. 
3. The external tool that should be present on every platform is `git diff`! and could be chosen as the default for this feature.  Following the inputs from @andymass, the command `git diff --ignore-all-spaces --color-words` works (I have tested it, and it is as good as Google's matchpatch).  A user configuration variable could be provided if the user would like to use `cwdiff` or `wdiff` instead.
4. Finally show the diff as a split buffer (maybe 80% vertical height of screen) with a virtual `left` (current) and `right` (last commit) view,  and place the cursor at the cursor location of the line being inserted. Of course, this being a hard line-wrap, `do`, `dp` etc. won't work, but this being a standard buffer, all yank/put operations shall.
5. Any changes made to the `left view upon closing the ephemeral diff buffer shall be updated into the document buffer.

Imagine this to be particularly powerful in the following scenario.
* Document author saves his text at the end of the day, commit it and goes home
* Next day fresh at work, the paragraph doesn't feel quite polished. The author reworks large chunks of the paragraph
* But suppose the author wants to refer to how much of an improvement was it over the previous night's work.
* They just select the paragraph in visual mode `vip` and press `\ld`
* A buffer with the paragraph's changes show up, making life easy for the author to know how much of an improvement has been achieved over last nights progress.

@all, please feel free to modify and continuously refine the workflow until something workable is in sight.

**Comment by lervag** [comment]
@krishnakumarg1984 Are you aware of [fugitive](https://github.com/tpope/vim-fugitive)? With it, you can do e.g. `:Gdiff`to open a git diff of your current file, or `:Gvdiff` for a vertical split. It seems to do almost exactly what you ask for. I personally add some convenient mappings like this:

```vim
nnoremap <silent><leader>ge :Gedit<cr>
nnoremap <silent><leader>gs :Gstatus<cr>
nnoremap <silent><leader>gd :Gdiff<cr>
```

The drawback, of course, is that this is not aware of any semantics. You can also use something like `:Git diff --word-diff`, which will open a buffer that shows the output, but I don't think this is as convenient. In any case, I personally find the fugitive method to work very well, although I would also agree it is not perfect.

For me, it feels very out of scope to implement a feature such as you ask. It feels much more as a separate plugin or as something that fits with a Git plugin or VCS plugin, like fugitive.

**Comment by lervag** [comment]
I'm closing this due to lack of activity. Although, as I mentioned, I think the features suggested here are better handled by e.g. fugitive.vim.
